### PR TITLE
refactor: improve GraphSelectionModel implementation, jsdoc and tests

### PR DESCRIPTION
**Pull request by redfish4ktc**
## Why

`GraphSelectionModel` and `SelectionChange` had no unit-test coverage despite selection being central to interaction and undo. The implementation also mixed legacy imperative loops with subtle quirks (a `[null]` guard that only inspected position 0, parameter reassignment, copy-paste JSDoc on `SelectionChange`) — the kind of code that's risky to change without a safety net.

### Tests
- **44+ unit tests** covering the public API of both classes, edge cases (empty arrays, null entries, single-selection mode, swap/round-trip on `SelectionChange`) and the `UNDO`/`CHANGE` event payloads. 100% branch coverage on both files.
- A `captureSelectionEvents` helper listens to both `InternalEvent.UNDO` and `InternalEvent.CHANGE` and returns `{ undo, change }`. Every UNDO-count assertion is paired with a CHANGE-count assertion, validating that the chain `changeSelection -> SelectionChange.execute -> fireEvent(CHANGE)` stays connected at the integration level.
- A regression sentinel under `addCell and addCells` documents a pre-existing bug uncovered during review (see _Known issue_ below). The tests assert the CURRENT buggy behavior so the suite stays green; they are tagged `BUG (#1055):` so the follow-up PR can find them.

### Bug fix
`changeSelection` now filters null/undefined entries from both arrays before applying. The previous guard `(added && added.length > 0 && added[0])` only inspected position 0, so `[realCell, null]` silently leaked `null` into the selection state. The new filter is strictly stronger and rejects nulls at any position.

### Implementation cleanup
- Imperative `for`-loops replaced with `filter`/`find`.
- `setCells` and `addCells` now share the same early-return shape, with no parameter reassignment and no redundant `isCellSelectable` calls.
- Every method declares an explicit return type.

### Documentation
- Rewrite `SelectionChange` JSDoc (was copy-pasted from `CurrentRootChange` and described root changes).
- Fix broken `{@link change}` / `{@link select}` references in `GraphSelectionModel` to point at the actual fired event (`InternalEvent.CHANGE`).

### Project rules
Introduce a **TypeScript** section in `coding-practices.md` grouping the existing _Type-Only Imports_ rule with a new _Explicit Return Types_ rule (improves IDE introspection and speeds up TS inference). Both rules are applied throughout the changed files.

## Known issue (not fixed here)

In single-selection mode, `addCells([alreadySelected])` (or with the already-selected cell as the first selectable candidate) clears the selection instead of being a no-op. Asserted in this PR via regression sentinels (tagged `BUG (#1055):`) so a follow-up PR can flip the assertions when fixing both `setCells` and `addCells`. See the [CodeRabbit review comment](https://github.com/maxGraph/maxGraph/pull/1055#pullrequestreview-4217110916) for the original finding.

## Test plan

- [x] `npm test -w packages/core` — all green, 100% branch coverage on the two files
- [x] `npm run lint`
- [x] `npm run build -w packages/core`

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests covering selection behavior, undoable selection changes, event emission, deduplication, and regression coverage.

* **Refactor**
  * Modernized selection logic and clarified method signatures to improve selection reliability and maintainability.

* **Chores**
  * Updated architecture coding-practices with TypeScript rules (type-only imports and explicit return types) and clarified selection-change documentation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/maxGraph/maxGraph/pull/1055?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/src/view/GraphSelectionModel.ts [hunk 1]
```diff
@@ -134,114 +135,97 @@ class GraphSelectionModel extends EventSource {
...
-        tmp.push(cells[i]);
-      }
-    }
-
-    this.changeSelection(tmp, remove);
+    const toAdd = cells.filter(
+      (cell) => !this.isSelected(cell) && this.graph.isCellSelectable(cell)
+    );
+    this.changeSelection(toAdd, null);
   }
```
[the remark this answers is not in the document]
Comment (redfish4ktc) [reply packages/core/src/view/GraphSelectionModel.ts:195]: This problem previously existed and need to be investigate.
Implementing this in addCells introduce a asymmetry with the setCells method which may also require the same change.
This will eventually managed in another PR

[the remark this answers is not in the document]
Comment (redfish4ktc) [reply packages/core/src/view/GraphSelectionModel.ts:195]: yes do it
